code/code/misc/info.cc

```diff
@@ -5385,642 +5385,291 @@ void TBeing::describeTrapDamType(const TTrap *obj, int) const
+      
+      if (includeClass && hasClass(classNumber) && spells.size() > 0) {
+        if (countValidClasses() > 1 && (whichClasses.size() == 0 || whichClasses.size() > 1))
+          buffer += format("  %s:\n\r") % className;
+        bufferSpells(spells, spacing);
       }
-      // can't we say if !cd, continue here?
-      if (cd && !cd->ok_for_class && getSkillValue(i) <= 0) 
-        continue;
+    }
+  };
 
-      totalcharges = 0;
-        
-      for (l = 0; l < 7; l++) {
-        if (search[l].where && wizlevel >= search[l].wizlevel) {
-          totalcharges += findComponentCharges(search[l].where, i);
-        }
-      }
+  if ((showTargeted && !showUtility) || (showOffensive && !showNonTargeted) || showAll)
+    bufferSpellsForCategory(TARGETED_OFFENSIVE);
 
-      if ((getSkillValue(i) <= 0) &&
-          (!tmp_var || (discArray[i]->start - tmp_var) > 0)) {
-	
-        if (!showall) 
-          continue;
+  if ((showTargeted && !showOffensive) || (showUtility && !showNonTargeted) || showAll)
+    bufferSpellsForCategory(TARGETED_UTILITY);
 
-        sprintf(buf, "%s%-22.22s%s  (Learned: %s)",  cyan(), discArray[i]->name, norm(), skill_diff(discArray[i]->start - tmp_var));
-      } else if (discArray[i]->toggle && !hasQuestBit(discArray[i]->toggle)) {
-          if (!showall) 
-          continue;
+  if ((showNonTargeted && !showUtility) || (showOffensive && !showTargeted) || showAll)
+    bufferSpellsForCategory(NONTARGETED_OFFENSIVE);
 
-          sprintf(buf, "%s%-22.22s%s  (Learned: Quest)", cyan(), discArray[i]->name, norm());
-      } else { 
-        if (getMaxSkillValue(i) < MAX_SKILL_LEARNEDNESS) {
-          if (discArray[i]->startLearnDo > 0) {
-            sprintf(learnbuf, "%.9s/%.9s", how_good(getSkillValue(i)), how_good(getMaxSkillValue(i))+1);
-            sprintf(buf, "%s%-22.22s%s %-19.19s", cyan(), discArray[i]->name, norm(), learnbuf);
-          } else 
-            sprintf(buf, "%s%-22.22s%s %-19.19s", cyan(), discArray[i]->name, norm(), how_good(getSkillValue(i)));   
-        } else 
-          sprintf(buf, "%s%-22.22s%s %-19.19s", cyan(), discArray[i]->name, norm(), how_good(getSkillValue(i)));
-            
-        unsigned int comp;
-
-        for (comp = 0; (comp < CompInfo.size()) && (i != CompInfo[comp].spell_num);comp++);
-
-        if (comp != CompInfo.size() && CompInfo[comp].comp_num >= 0) 
-          sprintf(buf + strlen(buf), "   [%2i] %s",  totalcharges, obj_index[real_object(CompInfo[comp].comp_num)].short_desc); 
-      }
-      strcat(buf, "\n\r");
-          
-      if (strlen(buf) + strlen(buffer) > (MAX_STRING_LENGTH * 2) - 2)
-        break;
+  if ((showNonTargeted && !showOffensive) || (showUtility && !showTargeted) || showAll)
+    bufferSpellsForCategory(NONTARGETED_UTILITY);
 
-      strcat(buffer, buf);
-    } 
-  }
-  d->page_string(buffer);
-}
+  if (!buffer.empty())
```

Comment:
Can it happen that a player types `spells` and nothing at all gets printed?

Reply:
Hmm, probably not actually, since it's always going to have at least a class/category header.

Fixed in https://github.com/sneezymud/sneezymud/pull/387/commits/1f8d735075569f6bcec13e15ea8f8091768b6bd3